File: R/packages/mops/src/hydrographFilter.f90
module sub
contains

subroutine heapsort(arr)
implicit none
!intent(out)
real(8),dimension(:),intent(inout):: arr
!local
real(8):: tmp
integer(4):: i,n
!code
n=size(arr)
do i=(n/2),1,-1
  CALL sift_down(i,n)
end do
do i=n,2,-1
  tmp= arr(1)
  arr(1)= arr(i)
  arr(i)= tmp
  CALL sift_down(1,i-1)
end do
contains
  subroutine sift_down(l,r)
  implicit none
  integer(4),intent(in):: l,r
  integer(4):: j,jold
  real(8):: a
  a=arr(l)
  jold=l
  j=l+l
  do
    if (j .gt. r) then
      exit
    end if
    if (j .lt. r) then
      if (arr(j) .lt. arr(j+1)) then
        j=j+1
      end if
    end if
    if (a .ge. arr(j)) then
      exit
    end if
    arr(jold)=arr(j)
    jold=j
    j=j+j
  end do
  arr(jold)=a
  end subroutine sift_down
end subroutine heapsort


function quantile(array,alpha) result (res)
implicit none
!intent(in)
real(8),dimension(:),intent(in):: array
real(8),intent(in):: alpha
!local
real(8):: res
real(8),dimension(size(array)):: array_sorted
real(8):: pos,posfrac,lower,upper
real(8),parameter:: ZERO= 0.0_8, ONE= 1.0_8
!code
if (alpha .le. ZERO) then
  res= minval(array)
else if (alpha .ge. ONE) then
  res= maxval(array)
else
  array_sorted(:)= array(:)
  call heapsort(array_sorted(:))
  if (size(array_sorted) .eq. 1) then
    res= array_sorted(1)
  else
    pos = ONE + alpha * dble(size(array_sorted) - 1) !Alpha=0 --> pos=1, Alpha=1 --> pos=n
    posfrac= pos - floor(pos)                        !Fractional part of pos
    lower = array_sorted(floor(pos))
    upper = array_sorted(floor(pos)+1)
    res= lower + posfrac * (upper - lower)
  end if
end if
end function quantile

end module sub

!-------------------------------------------------------------------------------

subroutine selectData(vect, vect_len, vect_na, win_len, win_lag, alpha, margins_ok, selected)
  use sub
  implicit none
  ! Intent(in)
  ! Vector with data
  real(8),dimension(vect_len),intent(in):: vect
  integer,intent(in):: vect_len
  real(8),intent(in):: vect_na
  ! Window specification
  integer,intent(in):: win_len
  integer,intent(in):: win_lag
  ! Selection threshold
  real(8),intent(in):: alpha
  ! Is selectio at window margins ok?
  logical,intent(in):: margins_ok
  ! Intent(out)
  logical,dimension(vect_len),intent(out):: selected
  ! Local
  integer:: mini, maxi, i
  logical,dimension(win_len):: local
  ! Code
  ! Initialize result
  selected= .false.
  ! Check errors
  if (win_lag .ge. 1) then
    ! Initialize indices
    mini= 1
    maxi= win_len
    ! Move through vector
    do while (maxi .le. vect_len)
      ! Initialize local window
      local= .false.
      ! Select data
      if (.not. any(vect(mini:maxi) .eq. vect_na)) then
        where(vect(mini:maxi) .gt. quantile(vect(mini:maxi), alpha)) local= .true.
      end if
      ! Undo selection at margins if requested
      if (.not. margins_ok) then
        do i=1,win_len
          if (.not. local(i)) exit
          local(i)= .false.
        end do
        do i=win_len,1,-1
          if (.not. local(i)) exit
          local(i)= .false.
        end do      
      end if
      ! Save result for local window (without unselecting formerly selected data)
      if (any(local)) where (local) selected(mini:maxi)= .true.
      ! Update indices
      mini= mini + win_lag
      maxi= maxi + win_lag
    end do
  end if
end subroutine selectData

File: R/packages/mops/R/hydrographFilter.r
#' Filter a hydrograph for periods of high or low flow
#'
#' This method allows a hydrograph to be filtered for flood or drought events.
#' Flood (drought) events are those where the observed variable exceeds (falls
#' below) a threshold value. The threshold value is conveniently defined as a
#' quantile. Furthermore, the threshold is variable in time. This is achieved by
#' applying the threshold-based selection concept to a moving window being
#' shifted along the time axis. The quantile (alpha value) and the parameters of
#' the moving windows are user-specified.
#'
#' @param vect Numeric vector representing the hydrograph, possibly containing
#'   \code{NA}s (see notes). It is assumed that the values represent observations
#'   taken at equal time intervals (regular time series).
#' @param windowLength Integer constant defining the width of the moving window
#'   (as number of observations) used in the data selection algorithm. Must be
#'   >= 1 and <= the length of \code{vect}. The window should be wide enough, i.e.
#'   contain a reasonable number of values, to estimate the desired quantile.
#' @param windowLag Integer constant defining the lag (as number of observations)
#'  being used when moving window along the time axis. Must be >= 1 and
#'  <= the length of \code{vect}. Typically, \code{windowLag} is chosen to be
#'  smaller than \code{windowLength} which results in an overlap of subsequent
#'  sampling windows. The computation time increases as \code{windowLag} decreases.
#' @param alpha A numeric value between 0 and 1. A value of 0.75 results in the
#'   selection of values exceeding the 'within-window' 75 percentile, for example.
#' @param marginsOK If \code{FALSE} (default), the algorithm only selects values
#'   in the center part of the moving window but not at the margins.
#'
#' @return A logical vector of the same length as \code{vect}. Elements are
#'   \code{TRUE} where the corresponding values in \code{vect} exceed the time-
#'   variable threshold defined by \code{alpha}. See the examples for how this
#'   vector can be used to filter \code{vect} for periods of floods or droughts. 
#'
#' @note Data will be selected only from those time windows where all
#'   corresponding elements in \code{vect} are finite according to
#'   \code{is.finite}. The selection algorithm tests for exceedence of threshold
#'   using the > operator (not >=). Thus, \code{TRUE} elements in the returned vector are
#'   \emph{greater than} the threshold while \code{FALSE} elements are \emph{equal
#'   or less than} the threshold.
#'
#' @author David Kneis \email{david.kneis@@uni-potsdam.de}
#'
#' @export
#' @useDynLib mops
#'
#' @examples
#' hydrograph= sin(seq(from=-pi/2, to=8*pi, by=0.05))+1
#' select_1= hydrographFilter(hydrograph, windowLag=1, alpha=0.8)
#' select_2= hydrographFilter(hydrograph, windowLag=1, alpha=0.2)
#' times= 1:length(hydrograph)
#' plot(x=times, y=hydrograph, xlab="Index", ylab="Data")
#' points(x=times[select_1],y=hydrograph[select_1], col="blue", pch=20)
#' points(x=times[!select_2],y=hydrograph[!select_2], col="red", pch=20)
#' legend("topright",pch=c(1,20,20),col=c("black","blue","red"),
#'   legend=c("Data","Floods","Droughts"))

hydrographFilter = function (
  vect,
  windowLength=ceiling(length(vect)/3),
  windowLag=ceiling(windowLength/10),
  alpha= 0.9,
  marginsOK= FALSE
) {
  # Check args
  if ((length(vect) == 0) || (!is.numeric(vect)))
    stop("Argument 'vect' must be numeric vector.")
  if ((length(windowLength) != 1) || (!is.numeric(windowLength)))
    stop("Argument 'windowLength' must be scalar integer.")
  if ((windowLength < 1) || (windowLength > length(vect)))
    stop("Expecting 1 <= 'windowLength' <= length of 'vect'.")
  if ((length(windowLag) != 1) || (!is.numeric(windowLag)))
    stop("Argument 'windowLag' must be scalar integer.")
  if ((windowLag < 1) || (windowLag > length(vect)))
    stop("Expecting 1 <= 'windowLag' <= length of 'vect'.")
  if ((length(alpha) != 1) || (!is.numeric(alpha)))
    stop("Argument 'alpha' must be scalar numeric value.")
  if ((alpha < 0) || (alpha > 1))
    stop("Expecting value in range (0,1) for argument 'alpha'.")
  if ((length(marginsOK) != 1) || (!is.logical(marginsOK)))
    stop("Argument 'marginsOK' must be scalar logical value.")
  # Convert NA values to a special numeric value
  nodata= min(vect,na.rm=TRUE) - 1.e06
  vect[!is.finite(vect)]= nodata
  # Call the fortran subroutine
  out= .Fortran("selectData",
    vect=         as.double(vect),
    vect_len=     as.integer(length(vect)),
    vect_na=      as.double(nodata),
    win_len=      as.integer(windowLength),
    win_lag=      as.integer(windowLag),
    alpha=        as.double(alpha),
    marginals_ok= as.integer(marginsOK),
    selected=     rep(as.integer(0),length(vect))
  ) 
  return(as.logical(out$selected))
}

